File: js/ValidTiles.js
import FieldSideDecorator from "./FieldSideDecorator";
import RoadSideDecorator from "./RoadSideDecorator";
import CitySideDecorator from "./CitySideDecorator";
import Tile from "./Tile";

export default class ValidTiles {
    constructor() {
        // Create SideDecorators for use in initialization.
        this.field_ = new FieldSideDecorator();
        this.road_ = new RoadSideDecorator();
        this.city_ = new CitySideDecorator();
        this.city_road_field_road_ = new Tile(this.city_, this.road_, this.field_, this.road_);
        this.city_road_field_road_.ConnectedRightToLeft();
        this.city_city_field_city_ = new Tile(this.city_, this.city_, this.field_, this.city_);
        this.city_city_field_city_.ConnectedTopToRight()
            .ConnectedTopToLeft()
            .ConnectedRightToLeft();
        // Add Pennant
        this.city_city_field_city_p_ = new Tile(this.city_city_field_city_);
        // Add Pennant
        this.city_city_city_city_p_ = new Tile(this.city_, this.city_, this.city_, this.city_);
        this.city_city_city_city_p_.ConnectedTopToRight()
            .ConnectedTopToBottom()
            .ConnectedTopToLeft()
            .ConnectedRightToBottom()
            .ConnectedRightToLeft()
            .ConnectedBottomToLeft();

    }

    city_road_field_road_copy() {
        return new Tile(this.city_road_field_road_);
    }

    city_city_field_city_copy() {
        return new Tile(this.city_city_field_city_);

    }

    city_city_field_city_p_copy() {
        return new Tile(this.city_city_field_city_p_);

    }

    city_city_city_city_p_copy() {
        return new Tile(this.city_city_city_city_p_);

    }
}

File: carcassonne-master/src/valid_tiles.cc
#include "valid_tiles.h"

#include <iostream>

#include "field_side_decorator.h"
#include "road_side_decorator.h"
#include "city_side_decorator.h"

namespace carcassonne
{

ValidTiles::ValidTiles()
{
	// Create SideDecorators for use in initialization.
	FieldSideDecorator field_;
 	RoadSideDecorator road_;
 	CitySideDecorator city_;
 	
	city_road_field_road_ = Tile(city_, road_, field_, road_);
	city_road_field_road_.ConnectedRightToLeft();

 	city_city_field_city_ = Tile(city_, city_, field_, city_);
 	city_city_field_city_.ConnectedTopToRight()
											 .ConnectedTopToLeft()
											 .ConnectedRightToLeft();

	// Add Pennant											
	city_city_field_city_p_ = Tile(city_city_field_city_);
												
	
	// Add Pennant
 	city_city_city_city_p_ = Tile(city_, city_, city_, city_);
 	city_city_city_city_p_.ConnectedTopToRight()
												.ConnectedTopToBottom()
												.ConnectedTopToLeft()
												.ConnectedRightToBottom()
												.ConnectedRightToLeft()
												.ConnectedBottomToLeft();
												 
}

Tile ValidTiles::city_road_field_road_copy()
{
	return Tile(city_road_field_road_);
}

Tile ValidTiles::city_city_field_city_copy()
{
	return Tile(city_city_field_city_);
	
}

Tile ValidTiles::city_city_field_city_p_copy()
{
	return Tile(city_city_field_city_p_);
	
}

Tile ValidTiles::city_city_city_city_p_copy()
{
	return Tile(city_city_city_city_p_);
	
}

ValidTiles::~ValidTiles()
{
}

}
